Add fractribution code (#1)
snowplow/dbt-snowplow-fractribution: README.md

![snowplow-logo](https://raw.githubusercontent.com/snowplow/dbt-snowplow-utils/main/assets/snowplow_logo.png)

# snowplow-fractribution

This dbt package:
- Uses page view and conversion events to perform fractional marketing attribution (fractribution) on your Snowplow data.
- Is used in conjunction with a Python script or Docker image to create the final output table.
- Is designed to be customized, allowing you to easily make modifications to suit your data and objectives. 

Please refer to the [doc site](https://docs.snowplow.io/docs/modeling-your-data/modeling-your-data-with-dbt/) for a full breakdown of the package. To run this package with detailed instructions and example data, see the [fractribution accelerator](https://docs.snowplow.io/accelerators/snowplow_fractribution/).

> ### Adaptor Support

Reviewer: ```suggestion
### Adapter Support
```

The snowplow-fractribution v0.1.0 package currently supports Snowflake. 

|      Warehouse     |    dbt versions     | snowplow-fractribution version |
| :----------------: | :-----------------: | :----------------------------: |
|      Snowflake     |  >=1.0.0 to <2.0.0  |             0.1.0              |

### Requirements

> - A dataset of web events from the [Snowplow Javascript Tracker](https://docs.snowplow.io/docs/collecting-data/collecting-from-own-applications/) and familiarity with the [snowplow-web](https://hub.getdbt.com/snowplow/snowplow_web/latest/) dbt package

Reviewer: ```suggestion
- A dataset of web events from the [Snowplow Javascript Tracker](https://docs.snowplow.io/docs/collecting-data/collecting-from-own-applications/) and familiarity with the [snowplow-web](https://hub.getdbt.com/snowplow/snowplow_web/latest/) dbt package
- The `snowplow_web_page_views` table generated by the snowplow-web package available in your warehouse
- A `conversions` dataset available in your warehouse (a dataset containing revenue data by users).
- A `channel spend` dataset available in your warehouse (a dataset containing total spend by channel)
```

The tables produced by the Python script are the following:

| Table                                      | Description                                                                           |
| ------------------------------------------ | ------------------------------------------------------------------------------------- |
| snowplow_fractribution_report_table        | The main output table that shows conversions, revenue, spend and ROAS per channel.
| snowplow_fractribution_channel_attribution | The conversion and revenue attribution per channel (used to create the report table).
| snowplow_fractribution_path_summary_with_channels | For each unique path, a summary of conversions, non conversions and revenue, as well as which channels were assigned the contribution |

### Setup steps

1. Configure the `conversion_clause` macro to filter your raw Snowplow events to successful conversion events.
2. Configure the `conversion_value` macro to return the value of the conversion event.
3. Configure the default `channel_classification` macro to yield your expected channels. The ROAS calculations / attribution calculations will run against these channel definitions.

> 4. Configure environment variables to be used by the `connection_parameters` in the `main_snowplow_snowflake.py` file to connect to Snowflake.

Reviewer: ```suggestion
4. Configure environment variables to be used by the `connection_parameters` in the `main_snowplow_snowflake.py` file to connect to Snowflake.
5. Overwrite default variables provided by the package in your dbt_project.yml file, if necessary. E.g.: make sure your `page_views_source` and `conversions_source` are aligned to what is available in your warehouse.
```

### Running

1. Ensure the setup steps have been completed above.
2. Run `dbt run`, or `dbt run --select package:fractribution`
3. Run the python script, e.g. `python utils/main_snowplow_snowflake.py --conversion_window_start_date '2022-06-03' --conversion_window_end_date '2022-08-01'` (see [doc site](https://docs.snowplow.io/docs/modeling-your-data/modeling-your-data-with-dbt/) for further details)

> ### Differences to Fractribution

Reviewer: ```suggestion
### Differences to Google's Fractribution
```

> There are some changes from the [original](https://github.com/google/fractribution) Fractribution code that have been noted below.

Reviewer: ```suggestion
There are some changes from [Google's](https://github.com/google/fractribution) Fractribution code that have been noted below.
```

Reviewer: As per recommendation on the Docs' PR.

Snowplow Fractribution (marketing attribution) model for dbt

### Setup steps

1. Configure the `conversion_clause` macro to filter your raw Snowplow events to successful conversion events.
2. Configure the `conversion_value` macro to return the value of the conversion event.
3. Configure the default `channel_classification` macro to yield your expected channels. The ROAS calculations / attribution calculations will run against these channel definitions.

### Running

1. Ensure the setup steps have been completed above.
2. Run `dbt run`, or `dbt run --select package:fractribution`

> 

Reviewer: Can we have an Adapter support section as well here, please? It is important for users to know it is for Snowflake and for which version etc. See example, [here](https://github.com/snowplow/dbt-snowplow-web/blob/main/README.md).

Author: Added, and also updated the Readme to make it more like the snowplow-web one. Let me know if I should change anything
